### PR TITLE
Add support for Http File Upload(XEP-0363)

### LINKED ISSUE #349: Support XEP-0363 HTTP_Upload
http_upload is supported by Conversations in Android, Monal in iOS and OSX as well as Gajim for Linux and is proposed standard xep-0363.

This is a very useful XEP for asynchronous picture sharing (Send picture while receiver is offline)

I would like to deploy a web chat for our users, but I have currently found no web client supporting this XEP.

Does Jsxc team plan to support http upload in the future?
##

### PR DESCRIPTION
**Pull request by sngyai**
Add Support for Http File Upload(XEP-0363). It is needed to modify your server to enable cross-domain.

fix #349

### CODE REVIEW

**Comment by sngyai** [comment]
Add Support for Http File Upload(XEP-0363). It is needed to modify your server to enable cross-domain.

**Comment by sngyai** [comment]
I am so sorry, I haven't created pull requests before

**Comment by tristan-k** [comment]
Anything which keeps this pull request from being merged into master?

**Comment by MarcelWaldvogel** [comment]
@sualko has been unavailable the past couple of months. Expect this to change soon.

**Comment by tristan-k** [comment]
Trying to apply the changes manually without rebuilding, but I dont know where I have to put [https://github.com/sngyai/jsxc/blob/20f23a181a268c37cf862502ca34ebfc38e22b61/src/jsxc.lib.httpupload.js](jsxc.lib.httpupload.js). Does it go into `ojsxc/js/jsxc/lib`?

File: src/jsxc.lib.httpupload.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                              size: file.size,
+                              type: file.type,
+                              data: (file.type.match(/^image\//)) ? img.attr('src') : null
+                           }
+                        });
+                        jsxc.gui.window.updateProgress(message, file.size, file.size);
+                        jsxc.xmpp.sendMessage(bid, putUrl, uid);
+                     }
+                  } else {
+                     jsxc.warn('ajax put failed');
```
Comment (sualko) [inline src/jsxc.lib.httpupload.js]: This message should be a little more detailed and specially catch the error conditions from http://xmpp.org/extensions/xep-0363.html#errors

File: src/jsxc.lib.httpupload.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                           dataType: 'text',
+                           attachment: {
+                              name: file.name,
+                              size: file.size,
+                              type: file.type,
+                              data: (file.type.match(/^image\//)) ? img.attr('src') : null
+                           }
+                        });
+                        jsxc.gui.window.updateProgress(message, file.size, file.size);
+                        jsxc.xmpp.sendMessage(bid, putUrl, uid);
```
Comment (sualko) [inline src/jsxc.lib.httpupload.js]: The get url should be used. Is there any XEP to "style" a download link?

File: src/jsxc.lib.httpupload.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+                           direction: 'out',
+                           dataType: 'text',
+                           attachment: {
+                              name: file.name,
+                              size: file.size,
+                              type: file.type,
+                              data: (file.type.match(/^image\//)) ? img.attr('src') : null
+                           }
+                        });
+                        jsxc.gui.window.updateProgress(message, file.size, file.size);
```
Comment (sualko) [inline src/jsxc.lib.httpupload.js]: This should be improved.

**Comment by sualko** [comment]
I think you copied some code from ``jsxc.muc``. That explains some of the stuff above... :laughing:

File: src/jsxc.lib.gui.js [hunk 4]
```diff
@@ -2050,7 +2050,12 @@ jsxc.gui.window = {
       win.find('.jsxc_sendFile').click(function() {
          $('body').click();
 
-         jsxc.gui.window.sendFile(bid);
+         if(typeof jsxc.options.get('httpupload') !== "undefined"){
```
Comment (DanScharon) [inline src/jsxc.lib.gui.js]: @sualko agreed, at least as long as the files and their filenames aren't stored encrypted on servers with http_upload support

**Comment by peterhoth** [comment]
ajax PUT in httpupload.sendfile will fail if the server implements self-certified SSL certificate. Tested with Chrome 53.0.2785.143 and Firefox 49.0.1. I have implemented XEP0363 in python and encountered this similar problem but i fixed it by passing a parameter to the PUT function: requests.put(... , verify=false). Is there a way to set ajax parameters to ignore untrusted site for PUT ?

**Comment by sualko** [comment]
@peterhoth Bypassing the internal browser certificate verification, would result in serious security issues. [Read also](http://stackoverflow.com/questions/25723415/bypassing-ssl-certificate-error-in-jquery-and-ajax). Nowadays everyone can gets a free certificate from e.g. https://letsencrypt.org/ and so self signed certificates don't have to be used anymore even for xmpp servers.

**Comment by sngyai** [comment]
Yes, I copied code from jsxc.muc because I'm new to javascript @sualko

**Comment by peterhoth** [comment]
@sualko, thanks. I managed to get the http_upload working and i am able to send a file to the server. However, i have encountered the following issues with the http_upload function.
1. Open chrome browser (Version 53.0.2785.143) and login to xmpp server
2. Clicked on user A to open a one to one chat window 
3. Sent a 7MB image to user (using http_upload) ---> Issue 1 : Does not show the upload progress. If it is a huge image, it will seem like the transfer is not taking place. 
4. Sent was successful with thumbnail image displayed in the chat window (Server contains the image file + thumbnail) ---> Issue 2 : How do i open the actual image ? Clicking on the thumbnail image does nothing.

**Comment by sualko** [comment]
@peterhoth This PR is still WIP and until it is merged, it will be further improved. Anyway thanks for your feedback.

@sngyai No problem. Are you interested in improving this PR further?

**Comment by sualko** [comment]
I close this issue in favor of #426

**Pull request closed without merging**